### PR TITLE
Feature/pdaf

### LINKED ISSUE #2: [TASK] PDAF
### Description of task

Probabilistic Data Association Filter

### Suggested Workflow

- [x] Look at earlier code by @hannahcl 
- [x] Read the Sensor Fusion Book
- [x] Use the already implemented EKF, MultiVarGauss, etc. already in the library

### Specifications

- [x] Output can be visualized

### Contacts

@chrstrom - Expert Target Tracker Person
@hannahcl - Wrote the Python PDAF
@eirikkolas, @alekskl01 - Are taking the TTK4250 Sensor Fusion course

Fundamentals of Sensor Fusion in `Teams -> Software -> Files -> Learning Materials -> Papers and other Resources`

### Code Quality

- [ ] Every function in header files are documented (inputs/returns/exceptions)
- [ ] The project has automated tests that cover MOST of the functions and branches in functions (pytest/gtest)
- [ ] The code is documented on the wiki (provide link)

### LINKED ISSUE #4: [TASK] IPDA
### Description of task

Extend the PDAF in #2 to also calculate the existence probability of the target

### Suggested Workflow

_No response_

### Specifications

_No response_

### Contacts

_No response_

### Code Quality

- [ ] Every function in header files are documented (inputs/returns/exceptions)
- [ ] The project has automated tests that cover MOST of the functions and branches in functions (pytest/gtest)
- [ ] The code is documented on the wiki (provide link)

### PR DESCRIPTION
**Pull request by TristanWolfram**
## Target tracking feature
* Implemented PDAF and IPDA with the use of vortex_filtering package (made by Eirik).
* This will be used for target tracking tasks.

### CODE REVIEW

File: vortex-filtering/include/vortex_filtering/filters/ipda.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    double prob_of_survival = 1.0;
+  };
+
+  /// @brief
+  /// @param measurements Measurements to iterate over
+  /// @param probability_of_survival How likely the object is to survive (Ps)
+  /// @param probability_of_detection How likely the object is to be detected (Pd)
+  /// @param clutter_intensity How likely it is to have a false positive
+  /// @param z_pred The predicted measurement
+  /// @return The existence probability
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/ipda.hpp]: Change to /** */ style comments as in all the other files
```suggestion
  /** 
   * @brief
   * @param measurements Measurements to iterate over
   * @param probability_of_survival How likely the object is to survive (Ps)
   * @param probability_of_detection How likely the object is to be detected (Pd)
   * @param clutter_intensity How likely it is to have a false positive
   * @param z_pred The predicted measurement
   * @return The existence probability
   */
```

File: vortex-filtering/include/vortex_filtering/filters/ipda.hpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    {
+      sum += z_pred.pdf(measurement);
+    }
+    double l_k = 1 - probability_of_detection + probability_of_detection / clutter_intensity * sum;
+
+    return (l_k * predicted_existence_probability) / (1 - (1 - l_k) * predicted_existence_probability);
+  }
+
+  static std::tuple<Gauss_x, double, std::vector<Vec_z>, std::vector<Vec_z>, Gauss_x, Gauss_z, std::vector<Gauss_x>>
+  step(const DynModT& dyn_model, const SensModT& sen_model, double timestep, const Gauss_x& x_est,
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/ipda.hpp:92]: Add doxygen

File: vortex-filtering/include/vortex_filtering/filters/ipda.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
+#pragma once
+#include <vector>
+#include <Eigen/Dense>
+#include <vortex_filtering/filters/pdaf.hpp>
+#include <vortex_filtering/models/dynamic_model_interfaces.hpp>
+#include <vortex_filtering/models/sensor_model_interfaces.hpp>
+
+namespace vortex::filter
+{
+template <concepts::DynamicModelLTVWithDefinedSizes DynModT, concepts::SensorModelLTVWithDefinedSizes SensModT> class IPDA {
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/ipda.hpp]: Add doxygen to class

File: vortex-filtering/include/vortex_filtering/filters/ipda.hpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  using T = Types_xzuvw<N_DIM_x, N_DIM_z, N_DIM_u, N_DIM_v, N_DIM_w>;
+
+  using EKF = vortex::filter::EKF_t<N_DIM_x, N_DIM_z, N_DIM_u, N_DIM_v, N_DIM_w>;
+
+  using Gauss_z  = typename T::Gauss_z;
+  using Gauss_x  = typename T::Gauss_x;
+  using Vec_z    = typename T::Vec_z;
+  using GaussMix = vortex::prob::GaussianMixture<N_DIM_x>;
+  using PDAF     = vortex::filter::PDAF<vortex::models::ConstantVelocity, vortex::models::IdentitySensorModel<4, 2>>;
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/ipda.hpp]: This should take the template parameters and not specific classes
```suggestion
  using PDAF     = vortex::filter::PDAF<DynModT, SensModT>;
```
The tests should be run with different models in order to catch this

File: vortex-filtering/include/vortex_filtering/filters/pdaf.hpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+#include <Eigen/Dense>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+#include <vortex_filtering/vortex_filtering.hpp>
+
+namespace vortex::filter {
+
+template <concepts::DynamicModelLTVWithDefinedSizes DynModT, concepts::SensorModelLTVWithDefinedSizes SensModT> class PDAF {
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/pdaf.hpp]: doxygen for class

File: vortex-filtering/include/vortex_filtering/filters/pdaf.hpp
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/pdaf.hpp:1]: Add doxygen comments for each member function

File: vortex-filtering/include/vortex_filtering/filters/pdaf.hpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+  using T = Types_xzuvw<N_DIM_x, N_DIM_z, N_DIM_u, N_DIM_v, N_DIM_w>;
+
+  using EKF            = vortex::filter::EKF_t<N_DIM_x, N_DIM_z, N_DIM_u, N_DIM_v, N_DIM_w>;
+  using Gauss_z        = typename T::Gauss_z;
+  using Gauss_x        = typename T::Gauss_x;
+  using Vec_z          = typename T::Vec_z;
+  using MeasurementsZd = std::vector<Vec_z>;
+  using StatesXd       = std::vector<Gauss_x>;
+  using GaussMixZd     = vortex::prob::GaussianMixture<N_DIM_x>;
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/pdaf.hpp]: Wrong dimension. Tests should catch this

```suggestion
  using GaussMix_z     = vortex::prob::GaussianMixture<N_DIM_z>;
```

File: vortex-filtering/include/vortex_filtering/filters/pdaf.hpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  static constexpr int N_DIM_w = SensModT::N_DIM_w;
+
+  using T = Types_xzuvw<N_DIM_x, N_DIM_z, N_DIM_u, N_DIM_v, N_DIM_w>;
+
+  using EKF            = vortex::filter::EKF_t<N_DIM_x, N_DIM_z, N_DIM_u, N_DIM_v, N_DIM_w>;
+  using Gauss_z        = typename T::Gauss_z;
+  using Gauss_x        = typename T::Gauss_x;
+  using Vec_z          = typename T::Vec_z;
+  using MeasurementsZd = std::vector<Vec_z>;
+  using StatesXd       = std::vector<Gauss_x>;
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/pdaf.hpp]: I think these typedefs are unnecessary.

File: vortex-filtering/test/ipda_test.cpp
Comment (EirikKolas) [inline vortex-filtering/test/ipda_test.cpp:1]: As mentioned in ipda.hpp, there should be at least one test where you use different models

File: vortex-filtering/test/pdaf_test.cpp
Comment (EirikKolas) [inline vortex-filtering/test/pdaf_test.cpp:1]: Most of these tests don't test anything, only plot

**Review by EirikKolas — COMMENTED** [review]
You should also update the readme in the filters folter with an explanation of how the PDAF and IPDA works

File: vortex-filtering/include/vortex_filtering/filters/README.md [hunk 8]
```diff
@@ -226,4 +228,91 @@ Vec2d z_meas{48, 65};
...
...mic model. We basically want to compare this model with the actual measurements and find a good estimated final value for the current state of the temperature in the drive. It's a bit of a design choice if you trust the model or the actual measurements. Both could be accurate or wrong. The main problem is that we can't put sensors directly into the drive (they would melt and be destroyed instantly). Therefore, we put a sensor outside the rocket and measure the temperature there. Here, we have to know how the temperature we measure corresponds to the actual temperature inside the drive. In other words, the temperature that we measure is not the actual value we need. But we know how to convert the measurement to the value we search for. This is the sensor model. If you want to get a better explanation of those models, I suggest using the Textbook "Artificial Intelligence: A Modern Approach, 4th Edition" by *Stuart Russell* and *Peter Norvig*. The book explains those models in Chapter 14 - *Probabilistic Reasoning Over Time*. All in all, you have to define those models and pass them to the function. Under the name space **vortex::models**, you can find simple predefined models to use:
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/README.md]: I think you mixed up the meaning of the sensor model here. The sensor model models what we would expect to measure given an estimated state (from a dynamic model). We don't convert the measurement to the value we search for, we convert the value we search for into the measurement

File: vortex-filtering/include/vortex_filtering/filters/README.md [hunk 9]
```diff
@@ -226,4 +228,91 @@ Vec2d z_meas{48, 65};
...
+* x estimation (the last known state of the target),
+* z measurements,
+* and config (which is a struct that holds:)
+    * mahalanobis threshold
+    * minimum gate threshold
+    * maximum gate threshold
+    * probability of detection
+    * clutter intensity
+
+#### Dynamic/Sensor Models
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/README.md:252]: Good explanation!

File: vortex-filtering/include/vortex_filtering/filters/README.md [hunk 10]
```diff
@@ -226,4 +228,91 @@ Vec2d z_meas{48, 65};
...
+// example how to use PDAF in pratice
+using PDAF = vortex::filter::PDAF<DynMod, SensorMod>;
+auto [x_final, inside, outside, x_pred, z_pred, x_updated] = PDAF::step(parameters...)
+```
+
+#### Previous State Estimate
+The step function needs to know what the previous state was. Based on this state, the dynamic model will be used. The model will give us a predicted new state. This state will be compared to the actual measurements. The previous state must be a Gaussian distribution. 
+
+#### Measurements
+The perceived measurements. This parameter consists of an Eigen vector. It should hold all perceived measurements. (The dimension must be the same as defined in the models.)
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/README.md]:
```suggestion
The perceived measurements. This parameter consists of an `Eigen::Vector`. It should hold all perceived measurements. (The dimension must be the same as defined in the models.)
```

File: vortex-filtering/include/vortex_filtering/filters/README.md [hunk 11]
```diff
@@ -226,4 +228,91 @@ Vec2d z_meas{48, 65};
...
+
+#### Basic Principle
+This will be a short description of what is happening inside the step function. For a more clear and specific explanation of the steps of the PDAF please look into the recommended textbooks.
+* The step function will first use the dynamic model to calculate the next state (only considering the given model). Then the sensor model is used to convert the predicted state into the measurement space. That means to get the value, we would measure with our sensor if the perceived state is like the predicted state (of the dynamic model). Concerning our rocket example: We predict the drive temperature with the dynamic model and use the sensor model to convert the drive temperature to the temperature we would measure from the outside.
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/README.md:274]: Nice!

File: vortex-filtering/include/vortex_filtering/filters/README.md [hunk 12]
```diff
@@ -226,4 +228,91 @@ Vec2d z_meas{48, 65};
...
+using ConstantVelocity = vortex::models::ConstantVelocity;
+using IdentitySensorModel = vortex::models::IdentitySensorModel<4, 2>;
+using IPDA = vortex::filter::IPDA<ConstantVelocity, IdentitySensorModel>;
+
+auto [x_final, existence_pred, inside, outside, x_pred, z_pred, x_updated] = IPDA::step(parameters...);
+```
+
+#### Parameters
+The parameters of the step() function are mainly the same as those of the PDAF. However, two parameters are added:
+* Survival estimate: This is the survival estimate of the previous state estimate (corresponding to x_estimate). It describes the target existence, which we also want to estimate for the current state.
```
Comment (EirikKolas) [inline vortex-filtering/include/vortex_filtering/filters/README.md]: existence probability?

File: vortex-filtering/include/vortex_filtering/filters/ipda.hpp (hunk 2 above)
Comment (EirikKolas) [reply vortex-filtering/include/vortex_filtering/filters/ipda.hpp:92]: Still needs doxygen

File: vortex-filtering/include/vortex_filtering/filters/pdaf.hpp
Comment (EirikKolas) [reply vortex-filtering/include/vortex_filtering/filters/pdaf.hpp:1]: Still needs doxygen comments

**Review by EirikKolas — COMMENTED** [review]
Nice readme. Please have a look through the rest of my comments from the previous review

File: vortex-filtering/include/vortex_filtering/filters/pdaf.hpp (hunk 7 above)
Comment (TristanWolfram) [reply vortex-filtering/include/vortex_filtering/filters/pdaf.hpp]: You are probably right, but I still think that StatesXd is easier to read in the actual code than the type definition.
So, they make the code easier to understand and therefore I would keep them.

File: vortex-filtering/include/vortex_filtering/filters/pdaf.hpp (hunk 6 above)
Comment (TristanWolfram) [reply vortex-filtering/include/vortex_filtering/filters/pdaf.hpp]: Yeah my bad. But the naming was wrong not the actual type. So, N_DIM_x is right.

**Review by EirikKolas — APPROVED** [review]
Noice!!

**Pull request merged**